### ISSUE TITLE
Clarify access to state variables via contract name

### ISSUE THREAD

**Issue by chriseth** [ethereum/solidity#988]
The following should either work (return 42) or be disallowed:

```
contract Scope {
    uint stateVar = 42;

    function getStateVar() constant returns (uint stateVar) {
        stateVar = Scope.stateVar;
    }
}
```

**Comment by axic** [comment]
As a note, calling functions actually works:

```
contract A {
    uint public x = 1;
    function A() {
        x = A.b();
    }
    function b() returns (uint) {
        return 3;
    }
}
```

`x` will be 3 after creation.

**Comment by axic** [comment]
@chriseth I think your initial example should be an error because using via `Scope` assumes it is an external call. Only accessing public variables should be allowed (i.e. an external call).

**Comment by axic** [comment]
```
contract A {
    uint x = 1;
    uint y = 2;
    function a() {
        x = A.y;
    }
}
```

results in

```
libsolidity/codegen/ContractCompiler.cpp(45): Throw in function void StackHeightChecker::check()
Dynamic exception type: boost::exception_detail::clone_impl<dev::solidity::InternalCompilerError>
std::exception::what: std::exception
[dev::tag_comment*] = I sense a disturbance in the stack.
```

and

```
contract A {
    uint x = 1;
    uint y = 2;
    function A() {
        x = A.y;
    }
}
```

results in

```
libsolidity/codegen/CompilerContext.h(57): Throw in function unsigned int dev::solidity::CompilerContext::stackHeight() const
Dynamic exception type: boost::exception_detail::clone_impl<dev::solidity::InternalCompilerError>
std::exception::what: std::exception
[dev::tag_comment*] = 
```

Same applies to `public` variables currently, though those should work.

**Comment by chriseth** [comment]
Scope access should always be internal access if the scope is a base contract or library and the identifier is an internal identifier (If the scope is not a base contract or library, access is restricted to types only).

To understand that, look at the following example:

```
contract A { function f() internal returns (uint) { return 1; } }
contract B is A {
  function f() internal returns (uint) { return 2; }
  function g() {
    A.f();
  }
}
```

I think the example illustrates that the possibility to provide internal access is necessary.

To understand that external access does not make sense, take into account that even if the `f` functions were marked `public`, it would not be possible to call `A.f` externally on the current contract, because the dispatcher will always go into `B.f`.

To summarize: All examples given in the issue description and comments should work, most of them are nonsensical, but they should work. We might disallow them on the ground that overloading state variables is not possible, but that is the only reason.

Another nice example that currently fails to compile properly:

```
pragma solidity ^0.4.0;

contract A {
    uint x = 7;
}
contract B {
    uint x = 9;
}
contract C is A, B {
    function f() returns (uint) {
        return A.x;
    }
}
```

**Comment by chriseth** [comment]
Fixed by #1264
